bson/marshal.go

```diff
@@ -171,7 +171,24 @@ var bufPool = sync.Pool{
 // See [Encoder] for more examples.
 func MarshalAppendWithContext(ec bsoncodec.EncodeContext, dst []byte, val interface{}) ([]byte, error) {
 	sw := bufPool.Get().(*bytes.Buffer)
-	defer bufPool.Put(sw)
+	defer func() {
+		// Proper usage of a sync.Pool requires each entry to have approximately
+		// the same memory cost. To obtain this property when the stored type
+		// contains a variably-sized buffer, we add a hard limit on the maximum
+		// buffer to place back in the pool. We limit the size to 16MiB because
+		// that's the maximum wire message size supported by any current MongoDB
+		// server.
+		//
+		// Comment based on
+		// https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/fmt/print.go;l=147
+		//
+		// Recycle byte slices that are smaller than 16MiB and at least half
+		// occupied.
+		if sw.Cap() < 16*1024*1024 && sw.Cap()/2 < sw.Len() {
+			bufPool.Put(sw)
+		}
+	}()
+
```

Comment:
This is normally best practice, but the Go standard library's "encoding/json" package does not perform this check when returning `*encodeState` to its pool ([encoding/json/encode.go#L268-L279](https://github.com/golang/go/blob/master/src/encoding/json/encode.go#L268-L279)). I assume the reasoning here is that a program that encodes large structs typically does so frequently. An example here would be a user that stores large documents in their Mongodb database - most often many of the blobs will be of similar size so if some are large - then many of the other documents will be large as well.

If we do keep this logic then we should probably only check the capacity since I'm not sure what the check for `sw.Cap()/2 < sw.Len()` buys us and may lead to thrashing when dealing with a mix of small and large BSON documents.

Reply:
Yep, the current "encoding/json" library doesn't do any pooled buffer size checking, which is actually the subject of https://github.com/golang/go/issues/27735. We got the idea to measure buffer utilization as a simple way to prevent holding onto a bunch of mostly unused memory from that issue.

Under "ideal" usage patterns where messages are all the same size, the pooled memory will be mostly utilized. The pathological case is when there are mostly small messages with a few very large messages. That can cause the pooled buffers to grow very large, holding onto a lot of memory, but also be underutilized by the small messages. That issue happened for some customers after we added buffer pooling to the operation logic and lead to [GODRIVER-2677](https://jira.mongodb.org/browse/GODRIVER-2677). The Go "encoding/json" library currently also suffers from the same memory pooling behavior.

One possible solution is to keep "buckets" of buffers of different size, so large buffers are used for large messages and small buffers are used for small messages. However, the implementation for that is relatively complex, so we decided to go with the simpler solution of setting a maximum buffer size and minimum utilization inspired by [this comment](https://github.com/golang/go/issues/27735#issuecomment-739169121). The trade off is holding onto less unused memory but possibly allocating memory more often for large messages.

Reply:
Thank you for pointing out that issue. I think the easiest safe thing to do would be to increase the size limit of the buffers that we'll pool to something like 32mb (2x the max size of BSON docs or even 24mb) and remove the check to see if half the capacity is used. The motivation here is that due to how `bytes.Buffer` grows a Buffer with a length over 8mb could easily have a cap greater than 16mb.

We should also check if the size is **equal to or less than** whatever limit we pick.. Also, this isn't that big of a deal since the difference here will likely only be an allocation or two.